File: static/js/staff.js
/**
 * Staff Management JavaScript
 * Handles staff CRUD operations and department assignments
 * Using DataTables for data display
 */

// Global variables
let staffTable;
let departments = [];
let currentStaffId = null;

$(document).ready(function() {
    // Show loading overlay
    $('#loading-overlay').show();
    
    // Initialize the staff table
    initStaffTable();
    
    // Load departments for dropdown
    loadDepartments();
    
    // Set up event listeners
    $('#add-staff-btn').on('click', showAddStaffModal);
    
    // Form submission handlers
    $('#staff-form').on('submit', function(e) {
        e.preventDefault();
        saveStaff();
    });
    
    $('#assignment-form').on('submit', function(e) {
        e.preventDefault();
        saveDepartmentAssignment();
    });
    
    // Delete confirmation handler
    $('#confirm-delete-btn').on('click', deleteStaff);
    
    // Table button handlers (delegated)
    $('#staff-table').on('click', '.btn-edit', function() {
        const staffId = $(this).data('staff-id');
        editStaff(staffId);
    });
    
    $('#staff-table').on('click', '.btn-delete', function() {
        const staffId = $(this).data('staff-id');
        confirmDelete(staffId);
    });
    
    $('#staff-table').on('click', '.btn-history', function() {
        const staffId = $(this).data('staff-id');
        viewHistory(staffId);
    });
    
    $('#staff-table').on('click', '.btn-assign', function() {
        const staffId = $(this).data('staff-id');
        showDepartmentAssignmentModal(staffId);
    });
});

/**
 * Initialize DataTables for displaying staff data
 */
function initStaffTable() {
    // Define position labels
    const positionLabels = {
        'NV': 'Nhân viên',
        'TK': 'Trưởng khoa',
        'PK': 'Phó khoa',
        'DDT': 'Điều dưỡng trưởng',
        'KTVT': 'Kỹ thuật viên trưởng',
        'KHAC': 'Khác'
    };
    
    // Initialize DataTable
    staffTable = $('#staff-table').DataTable({
        ajax: {
            url: "/api/staff",
            method: "GET",
            dataSrc: "staff"
        },
        columns: [
            { 
                data: "StaffId",
                title: "ID", 
                width: "5%"
            },
            { 
                data: "StaffName", 
                title: "Name",
                width: "20%"
            },
            { 
                data: "StaffRole", 
                title: "Role",
                width: "15%"
            },
            { 
                data: "DepartmentName", 
                title: "Department",
                width: "20%",
                render: function(data, type, row) {
                    return data || '<em class="text-muted">Not assigned</em>';
                }
            },
            { 
                data: "Position",
                title: "Position",
                width: "15%", 
                render: function(data, type, row) {
                    if (type === 'sort' || type === 'type') {
                        return data || '';
                    }
                    return data ? `${data} - ${positionLabels[data] || ''}` : 'N/A';
                }
            },
            { 
                data: "StaffAvailable", 
                title: "Availability",
                width: "10%",
                render: function(data, type, row) {
                    if (type === 'sort' || type === 'type') {
                        return data ? 1 : 0;
                    }
                    return data ? 
                        '<span class="badge bg-success">Available</span>' : 
                        '<span class="badge bg-danger">Not Available</span>';
                }
            },
            { 
                data: null, 
                title: "Actions",
                width: "15%",
                orderable: false,
                className: 'text-center',
                render: function(data, type, row) {
                    return `
                    <div class="btn-group btn-group-sm">
                        <button class="btn btn-outline-info btn-history" data-staff-id="${row.StaffId}" title="View History">
                            <i class="fas fa-history"></i>
                        </button>
                        <button class="btn btn-outline-primary btn-edit" data-staff-id="${row.StaffId}" title="Edit">
                            <i class="fas fa-edit"></i>
                        </button>
                        <button class="btn btn-outline-success btn-assign" data-staff-id="${row.StaffId}" title="Assign Department">
                            <i class="fas fa-hospital"></i>
                        </button>
                        <button class="btn btn-outline-danger btn-delete" data-staff-id="${row.StaffId}" title="Delete">
                            <i class="fas fa-trash"></i>
                        </button>
                    </div>`;
                }
            }
        ],
        responsive: true,
        lengthMenu: [5, 10, 25, 50],
        pageLength: 10,
        language: {
            emptyTable: "No staff records found",
            zeroRecords: "No matching staff records found"
        },
        drawCallback: function() {
            $('#loading-overlay').hide();
        }
    });
}

/**
 * Load departments for the dropdown menu
 */
function loadDepartments() {
    $.ajax({
        url: '/api/departments',
        method: 'GET',
        success: function(response) {
            departments = response.departments;
            const departmentSelect = $('#department-id');
            
            // Clear existing options
            departmentSelect.empty().append('<option value="">Select a department</option>');
            
            // Add department options
            departments.forEach(function(dept) {
                departmentSelect.append(
                    `<option value="${dept.DepartmentId}">${dept.DepartmentName} (${dept.DepartmentType || 'N/A'})</option>`
                );
            });
        },
        error: function(error) {
            console.error('Error loading departments:', error);
            showAlert('danger', 'Failed to load departments. Please refresh the page.');
            
            // Hide loading overlay if visible
            $('#loading-overlay').hide();
        }
    });
}

/**
 * Show the modal for adding a new staff member
 */
function showAddStaffModal() {
    currentStaffId = null;
    
    // Update modal title
    $('#staff-modal-title').text('Add New Staff');
    
    // Reset form
    $('#staff-form')[0].reset();
    $('#staff-id').val('');
    
    // Show department field for new staff
    $('#department-container').removeClass('d-none');
    
    // Show the modal
    $('#staff-modal').modal('show');
}

/**
 * Show the modal for editing an existing staff member
 */
function editStaff(staffId) {
    currentStaffId = staffId;
    
    // Show loading overlay
    $('#loading-overlay').show();
    
    // Fetch staff data
    $.ajax({
        url: `/api/staff/${staffId}`,
        method: 'GET',
        success: function(response) {
            const staff = response.staff;
            
            // Update modal title
            $('#staff-modal-title').text(`Edit Staff: ${staff.StaffName}`);
            
            // Populate form fields
            $('#staff-id').val(staff.StaffId);
            $('#staff-name').val(staff.StaffName);
            $('#staff-role').val(staff.StaffRole);
            $('#department-id').val(staff.DepartmentId || '');
            $('#position').val(staff.Position || '');
            $('#staff-available').prop('checked', staff.StaffAvailable);
            
            // For editing existing staff, hide department field as it should be managed through assignments
            if (staff.DepartmentId) {
                $('#current-department-text').text(staff.DepartmentName);
                $('#current-department-container').removeClass('d-none');
                $('#department-container').addClass('d-none');
            } else {
                $('#current-department-container').addClass('d-none');
                $('#department-container').removeClass('d-none');
            }
            
            // Hide loading overlay
            $('#loading-overlay').hide();
            
            // Show the modal
            $('#staff-modal').modal('show');
        },
        error: function(error) {
            console.error('Error fetching staff details:', error);
            showAlert('danger', 'Failed to load staff details. Please try again.');
            
            // Hide loading overlay
            $('#loading-overlay').hide();
        }
    });
}

/**
 * Save or update staff member
 */
function saveStaff() {
    // Validate form
    const form = $('#staff-form')[0];
    if (!form.checkValidity()) {
        form.reportValidity();
        return;
    }
    
    // Show loading overlay
    $('#loading-overlay').show();
    
    // Gather form data
    const staffData = {
        StaffName: $('#staff-name').val(),
        StaffRole: $('#staff-role').val(),
        StaffAvailable: $('#staff-available').is(':checked'),
    };
    
    // For new staff, include department information if provided
    if (!currentStaffId) {
        const departmentId = $('#department-id').val();
        const position = $('#position').val();
        
        if (departmentId) {
            staffData.DepartmentId = departmentId;
            staffData.Position = position || null;
        }
    }
    
    // Determine if creating or updating
    const isNewStaff = !currentStaffId;
    const url = isNewStaff ? '/api/staff' : `/api/staff/${currentStaffId}`;
    const method = isNewStaff ? 'POST' : 'PUT';
    
    // If updating, add staff ID to data
    if (!isNewStaff) {
        staffData.StaffId = currentStaffId;
    }
    
    // Send request
    $.ajax({
        url: url,
        method: method,
        contentType: 'application/json',
        data: JSON.stringify(staffData),
        success: function(response) {
            // Hide the modal
            $('#staff-modal').modal('hide');
            
            // Refresh the table
            staffTable.ajax.reload();
            
            // Show success message
            showAlert('success', isNewStaff ? 'Staff created successfully' : 'Staff updated successfully');
            
            // Hide loading overlay
            $('#loading-overlay').hide();
        },
        error: function(error) {
            console.error('Error saving staff:', error);
            showAlert('danger', `Failed to ${isNewStaff ? 'create' : 'update'} staff. Please try again.`);
            
            // Hide loading overlay
            $('#loading-overlay').hide();
        }
    });
}

/**
 * Show confirmation dialog for deleting staff
 */
function confirmDelete(staffId) {
    currentStaffId = staffId;
    
    // Load staff details for confirmation message
    $.ajax({
        url: `/api/staff/${staffId}`,
        method: 'GET',
        success: function(response) {
            const staff = response.staff;
            
            // Set confirmation message
            $('#delete-confirmation-message').html(
                `Are you sure you want to delete <strong>${staff.StaffName}</strong>? ` +
                `This action cannot be undone.`
            );
            
            // Show the confirmation modal
            $('#delete-modal').modal('show');
        },
        error: function(error) {
            console.error('Error loading staff details:', error);
            showAlert('danger', 'Failed to load staff details. Please try again.');
        }
    });
}

/**
 * Delete staff after confirmation
 */
function deleteStaff() {
    if (!currentStaffId) return;
    
    // Show loading overlay
    $('#loading-overlay').show();
    
    $.ajax({
        url: `/api/staff/${currentStaffId}`,
        method: 'DELETE',
        success: function(response) {
            // Hide the confirmation modal
            $('#delete-modal').modal('hide');
            
            // Refresh the table
            staffTable.ajax.reload();
            
            // Show success message
            showAlert('success', 'Staff deleted successfully');
            
            // Hide loading overlay
            $('#loading-overlay').hide();
        },
        error: function(error) {
            console.error('Error deleting staff:', error);
            
            // Hide the confirmation modal
            $('#delete-modal').modal('hide');
            
            // Show error message
            showAlert('danger', 'Failed to delete staff. Please try again.');
            
            // Hide loading overlay
            $('#loading-overlay').hide();
        }
    });
}

/**
 * Navigate to staff history page
 */
function viewHistory(staffId) {
    window.location.href = `/staff/history/${staffId}`;
}

/**
 * Show department assignment modal
 */
function showDepartmentAssignmentModal(staffId) {
    currentStaffId = staffId;
    
    // Reset form
    $('#assignment-form')[0].reset();
    
    // Load staff details
    showLoadingOverlay();
    
    // Load departments for dropdown
    $.ajax({
        url: '/api/departments',
        method: 'GET',
        success: function(response) {
            departments = response.departments;
            const departmentSelect = $('#department');
            
            // Clear existing options
            departmentSelect.empty().append('<option value="">Select a department</option>');
            
            // Add department options
            departments.forEach(function(dept) {
                departmentSelect.append(
                    `<option value="${dept.DepartmentId}">${dept.DepartmentName} (${dept.DepartmentType || 'N/A'})</option>`
                );
            });
            
            // Load staff details
            $.ajax({
                url: `/api/staff/${staffId}`,
                method: 'GET',
                success: function(response) {
                    const staff = response.staff;
                    
                    // Update modal title
                    $('#assignment-modal-title').text(`Assign Department: ${staff.StaffName}`);
                    
                    // Show current department if exists
                    if (staff.DepartmentId) {
                        $('#current-department').text(staff.DepartmentName);
                        $('#current-department-container').removeClass('d-none');
                        
                        // Pre-select current department in dropdown
                        $('#department').val(staff.DepartmentId);
                        $('#position').val(staff.Position || '');
                    } else {
                        $('#current-department-container').addClass('d-none');
                    }
                    
                    // Show modal
                    $('#assignment-modal').modal('show');
                    
                    hideLoadingOverlay();
                },
                error: function(error) {
                    console.error('Error loading staff:', error);
                    showAlert('danger', 'Failed to load staff data. Please try again.');
                    hideLoadingOverlay();
                }
            });
        },
        error: function(error) {
            console.error('Error loading departments:', error);
            showAlert('danger', 'Failed to load departments. Please try again.');
            hideLoadingOverlay();
        }
    });
}

/**
 * Save department assignment
 */
function saveDepartmentAssignment() {
    const departmentId = $('#department').val();
    const position = $('#position').val();
    
    if (!departmentId) {
        showAlert('warning', 'Please select a department.', 'assignment-alert');
        return;
    }
    
    const assignmentData = {
        DepartmentId: departmentId,
        Position: position
    };
    
    // Show loading overlay
    showLoadingOverlay();
    
    $.ajax({
        url: `/api/staff/${currentStaffId}/department`,
        method: 'POST',
        contentType: 'application/json',
        data: JSON.stringify(assignmentData),
        success: function(response) {
            // Hide modal
            $('#assignment-modal').modal('hide');
            
            // Refresh data table
            staffTable.ajax.reload();
            
            // Show success message
            showAlert('success', 'Department assignment updated successfully!');
            
            hideLoadingOverlay();
        },
        error: function(error) {
            console.error('Error assigning department:', error);
            showAlert('danger', 'Failed to update department assignment. Please try again.', 'assignment-alert');
            hideLoadingOverlay();
        }
    });
}

/**
 * Show alert message
 */
function showAlert(type, message, targetId = 'main-alert') {
    const alertElement = $(`#${targetId}`);
    
    // Set message and type
    alertElement.html(`
        <div class="alert alert-${type} alert-dismissible fade show" role="alert">
            ${message}
            <button type="button" class="btn-close" data-bs-dismiss="alert" aria-label="Close"></button>
        </div>
    `);
    
    // Show alert
    alertElement.removeClass('d-none');
    
    // Auto-hide after 5 seconds
    setTimeout(function() {
        alertElement.find('.alert').alert('close');
    }, 5000);
}

/**
 * Show loading overlay
 */
function showLoadingOverlay() {
    $('#loading-overlay').show();
}

/**
 * Hide loading overlay
 */
function hideLoadingOverlay() {
    $('#loading-overlay').hide();
}

File: api/staff.py
"""
Staff API endpoints
"""

from flask import Blueprint, jsonify, request
from models_main import db
from models.Staff import Staff
from models.Department import Department
from models.StaffDepartment import StaffDepartment
from sqlalchemy import desc
from datetime import datetime

staff_bp = Blueprint('staff_bp', __name__)

@staff_bp.route('/staff', methods=['GET'])
def get_all_staff():
    """Get all staff members with their current department"""
    try:
        # Get query parameters
        department_id = request.args.get('department_id', type=int)
        
        staff_list = Staff.query.all()
        result = []
        
        for staff in staff_list:
            # Only include available staff
            if not staff.StaffAvailable:
                continue
                
            staff_dict = staff.to_dict()
            current_dept_assignment = StaffDepartment.query.filter_by(
                StaffId=staff.StaffId, 
                Current=True
            ).first()
            
            if current_dept_assignment:
                department = Department.query.get(current_dept_assignment.DepartmentId)
                staff_dict['DepartmentId'] = department.DepartmentId
                staff_dict['DepartmentName'] = department.DepartmentName
                staff_dict['Position'] = current_dept_assignment.Position
                
                # Filter by department if specified
                if department_id and current_dept_assignment.DepartmentId != department_id:
                    continue
            else:
                staff_dict['DepartmentId'] = None
                staff_dict['DepartmentName'] = None
                staff_dict['Position'] = None
                
                # Skip staff without department if filtering by department
                continue
                
            result.append(staff_dict)
            
        return jsonify({'staff': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/staff/<int:staff_id>', methods=['GET'])
def get_staff(staff_id):
    """Get a specific staff member by ID"""
    try:
        staff = Staff.query.get(staff_id)
        if not staff:
            return jsonify({'error': 'Staff not found'}), 404
            
        staff_dict = staff.to_dict()
        current_dept_assignment = StaffDepartment.query.filter_by(
            StaffId=staff.StaffId, 
            Current=True
        ).first()
        
        if current_dept_assignment:
            department = Department.query.get(current_dept_assignment.DepartmentId)
            staff_dict['DepartmentId'] = department.DepartmentId
            staff_dict['DepartmentName'] = department.DepartmentName
            staff_dict['Position'] = current_dept_assignment.Position
        else:
            staff_dict['DepartmentId'] = None
            staff_dict['DepartmentName'] = None
            staff_dict['Position'] = None
            
        return jsonify({'staff': staff_dict})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/staff', methods=['POST'])
def create_staff():
    """Create a new staff member"""
    try:
        data = request.json
        
        # Check if staff already exists
        if Staff.query.get(data.get('StaffId')):
            return jsonify({'error': 'Staff ID already exists'}), 400
            
        # Create new staff
        new_staff = Staff(
            StaffId=data.get('StaffId'),
            StaffName=data.get('StaffName'),
            StaffRole=data.get('StaffRole'),
            StaffAvailable=data.get('StaffAvailable', True)
        )
        
        db.session.add(new_staff)
        
        # If department provided, create department assignment
        department_id = data.get('DepartmentId')
        if department_id:
            # First check if department exists
            dept = Department.query.get(department_id)
            if not dept:
                return jsonify({'error': 'Department not found'}), 400
                
            # Create department assignment
            dept_assignment = StaffDepartment(
                StaffId=new_staff.StaffId,
                DepartmentId=department_id,
                Current=True,
                Position=data.get('Position')
            )
            db.session.add(dept_assignment)
            
        db.session.commit()
        return jsonify({'message': 'Staff created successfully', 'staffId': new_staff.StaffId}), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/staff/<int:staff_id>', methods=['PUT'])
def update_staff(staff_id):
    """Update an existing staff member"""
    try:
        staff = Staff.query.get(staff_id)
        if not staff:
            return jsonify({'error': 'Staff not found'}), 404
            
        data = request.json
        
        # Update staff details
        if 'StaffName' in data:
            staff.StaffName = data['StaffName']
        if 'StaffRole' in data:
            staff.StaffRole = data['StaffRole']
        if 'StaffAvailable' in data:
            staff.StaffAvailable = data['StaffAvailable']
            
        # Handle department assignment
        if 'DepartmentId' in data:
            # Get current assignment if any
            current_assignment = StaffDepartment.query.filter_by(
                StaffId=staff_id,
                Current=True
            ).first()
            
            # If department changed or no current assignment
            if not current_assignment or current_assignment.DepartmentId != data['DepartmentId']:
                # Set current assignment to inactive if exists
                if current_assignment:
                    current_assignment.Current = False
                    
                # Create new assignment
                new_assignment = StaffDepartment(
                    StaffId=staff_id,
                    DepartmentId=data['DepartmentId'],
                    Current=True,
                    Position=data.get('Position', current_assignment.Position if current_assignment else None)
                )
                db.session.add(new_assignment)
            # If position changed but department is the same
            elif current_assignment and 'Position' in data and current_assignment.Position != data['Position']:
                current_assignment.Position = data['Position']
                
        db.session.commit()
        return jsonify({'message': 'Staff updated successfully'})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/staff/<int:staff_id>', methods=['DELETE'])
def delete_staff(staff_id):
    """Delete a staff member"""
    try:
        staff = Staff.query.get(staff_id)
        if not staff:
            return jsonify({'error': 'Staff not found'}), 404
            
        # Delete all department assignments
        StaffDepartment.query.filter_by(StaffId=staff_id).delete()
        
        # Delete the staff
        db.session.delete(staff)
        db.session.commit()
        
        return jsonify({'message': 'Staff deleted successfully'})
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/staff/<int:staff_id>/history', methods=['GET'])
def get_staff_history(staff_id):
    """Get department assignment history for a staff member"""
    try:
        # Check if staff exists
        staff = Staff.query.get(staff_id)
        if not staff:
            return jsonify({'error': 'Staff not found'}), 404
            
        # Get all department assignments for this staff
        assignments = StaffDepartment.query.filter_by(StaffId=staff_id).all()
        
        result = []
        for assignment in assignments:
            department = Department.query.get(assignment.DepartmentId)
            result.append({
                'StaffId': assignment.StaffId,
                'DepartmentId': assignment.DepartmentId,
                'DepartmentName': department.DepartmentName,
                'Current': assignment.Current,
                'Position': assignment.Position
            })
            
        return jsonify({'history': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@staff_bp.route('/departments', methods=['GET'])
def get_departments():
    """Get all departments for dropdown menus"""
    try:
        departments = Department.query.all()
        result = [dept.to_dict() for dept in departments]
        return jsonify({'departments': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500
@staff_bp.route('/staff/<int:staff_id>/department', methods=['POST'])
def assign_department(staff_id):
    """Assign staff to a department"""
    try:
        # Check if staff exists
        staff = Staff.query.get(staff_id)
        if not staff:
            return jsonify({'error': 'Staff not found'}), 404
            
        data = request.json
        department_id = data.get('DepartmentId')
        position = data.get('Position')
        
        # Check if department exists
        department = Department.query.get(department_id)
        if not department:
            return jsonify({'error': 'Department not found'}), 404
            
        # Get current assignment if any
        current_assignment = StaffDepartment.query.filter_by(
            StaffId=staff_id,
            Current=True
        ).first()
        
        # If there's a current assignment
        if current_assignment:
            # If department is the same, just update position if provided
            if current_assignment.DepartmentId == department_id:
                if position:
                    current_assignment.Position = position
                    
                db.session.commit()
                return jsonify({
                    'message': 'Department assignment updated successfully',
                    'departmentId': department_id,
                    'departmentName': department.DepartmentName
                })
            else:
                # Set current assignment to inactive
                current_assignment.Current = False
        
        # Create new department assignment
        new_assignment = StaffDepartment(
            StaffId=staff_id,
            DepartmentId=department_id,
            Position=position,
            Current=True
        )
        
        db.session.add(new_assignment)
        db.session.commit()
        
        return jsonify({
            'message': 'Department assignment updated successfully',
            'departmentId': department_id,
            'departmentName': department.DepartmentName
        })
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500
